docs/CEK.md

```diff
@@ -0,0 +1,200 @@
+# Two versions of the CEK machine
+
+
+Version 1 has values in the environment. Version 2 has closures (pairs
+of values and environments) in the environment.
+
+In my limited testing I have found version 1 to be faster and I think
+it is simpler.
+
+The main point of this document is so that we can discuss whether to
+use version 1 in the implementation. Version 2 is described as a
+comparison.
+
+## 1. Value version
+
+Note: that this makes no attempt to delay doing type
+substitutions. This would be a further optimisation and I think it
+could be modified to do so. To get an idea, and avoid this issue, I
+guess one can focus on the other parts of the syntax (constants,
+lambda, application, wrap/unwrap).
+
+Values and environments are mutually defined. Environments only appear
+under binders, they have entries for all free variables in the term
+but no entry for the variable (`α` or `x`) bound by the binder.
+
+```
+V := con cn
+   | abs ɑ (M , ⍴)
+   | lam x (M , ρ)
+   | iwrap A B V
+ρ := []
+   | ρ[x |-> V]`
+```
+
+Frames: note that we never have pairs of values and environments, only
+pairs of terms and environments in frames. In the builtin case, we
+stash the environment `ρ` that we were in when we began processing the
+builtin. We use ρ to evaluate each of the builtin arguments in,
+kicking off evaluation of each argument with the original ρ. The
+builtin frame refers to the point where we are processing the
+arguments from left to right, the already evaluated args are stored in
+`Vs` and the as yet unevaluated args are stored in `Ms`.
+
+```
+f := [_ (M,ρ) ]
```

Comment:
A possible simplification occurs to me. We have pairs of terms (not values!) and environments in a couple of places: they're stored in frames like this, they're stored in values, and they're also implicitly present in the computation rules.

So define `C` (for "computation", rather than closure) as `(M, p)`

Now we have:
```
V := ...
     | abs a C
     ...

f := [ _ C ]
  ...
```
And the compute rule operates on a "computation" (we move the environment from the "context" in to the "argument").

This seems pretty nice to me: all the places where `C` appears are "unfinished" computations of values, whereas `V` is a "finished" value.

Reply:
Those two `C`s are different, the one under a binder has one fewer entry than the corresponding term and the one in the frame has the same amount. It's easier to think of `lam` instead of `abs`, in `lam x (M,ρ)` - `ρ` has entries for all variables mentioned in `M` except for `x`. Evaluation is stuck. In the application frame, the environment `ρ` has entries for all variables in the accompanying argument term and evaluation is just paused.

Reply:
Yes, I got that: but they have the same type. Maybe that's a false friend, but it seems like it could be helpful.